common/src/main/java/com/microsoft/hydralab/common/management/listener/impl/PreInstallListener.java

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+package com.microsoft.hydralab.common.management.listener.impl;
+
+import com.android.ddmlib.InstallException;
+import com.microsoft.hydralab.common.entity.common.DeviceInfo;
+import com.microsoft.hydralab.common.management.DeviceManager;
+import com.microsoft.hydralab.common.management.listener.DeviceListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+
+/**
+ * @author zhoule
+ * @date 01/17/2023
+ */
+
+public class PreInstallListener implements DeviceListener {
+    DeviceManager deviceManager;
+    private Logger classLogger = LoggerFactory.getLogger(PreInstallListener.class);
+
+    public PreInstallListener(DeviceManager deviceManager) {
+        this.deviceManager = deviceManager;
+    }
+
+    @Override
+    public void onDeviceInactive(DeviceInfo deviceInfo) {
+
+    }
+
+    @Override
+    public void onDeviceConnected(DeviceInfo deviceInfo) {
+        File appDir = deviceManager.getPreAppDir();
+        File[] appFiles = appDir.listFiles();
+        for (File appFile : appFiles) {
+            if (appFile.isFile()) {
+                try {
+                    deviceManager.installApp(deviceInfo, appFile.getAbsolutePath(), classLogger);
+                } catch (InstallException e) {
+                    classLogger.error(String.format("Pre-Install %s failed", appFile.getAbsolutePath()), e);
+                }
```

Comment:
I like the idea of device status listener, several suggestions:
- Rename DeviceListener to DeviceStatusListener
- I think it's a better idea to rethrow this InstallException instead of eating it with logging. If we are going to fail, fail fast.
- Reduce the indents by "if not, continue".

Reply:
Have updated 1&3. About 2, this listener will be invoked when device connected, and can't stop test process or agent main process. So, I don't think we should throw this exception. It can't do anything but to block other listeners.

Reply:
Updated. Will exit if failed.